ultravox/training/train.py

```diff
@@ -170,21 +173,42 @@ def main() -> None:
             use_mds=args.mds,
             mds_batch_size=args.batch_size,
         )
-        data_sets = [datasets.create_dataset(ds, data_args) for ds in args.data_sets]
-        interleaved = datasets.InterleaveDataset(data_sets, repeat=args.repeat_data)
-        train_dataset: torch.utils.data.IterableDataset = (
-            ultravox_processing.UltravoxDataproc(
-                interleaved, processor=processor, train_on_inputs=args.train_on_inputs
-            )
+        val_data_args = datasets.VoiceDatasetArgs(
+            num_prompts=1,
+            data_dir=args.data_dir,
+            shuffle=False,
+            max_audio_duration_secs=16,
+            use_mds=args.mds,
+            mds_batch_size=args.batch_size,
         )
-        train_dataset = datasets.Range(train_dataset, args.num_samples)
+        train_dataset, val_dataset = [
+            datasets.Range(
+                ultravox_processing.UltravoxDataproc(
+                    datasets.InterleaveDataset(
+                        [
+                            datasets.create_dataset(ds, split_args)
+                            for ds in args.data_sets
+                        ],
+                        repeat=args.repeat_data,
+                    ),
+                    processor=processor,
+                    train_on_inputs=args.train_on_inputs,
+                ),
+                num_samples=num_samples,
+            )
+            for split_args, num_samples in [
+                (data_args, args.num_samples),
+                (val_data_args, args.val_num_samples),
+            ]
+        ]
```

Comment:
This looks ugly, I agree. Let me know if you have a better idea. Other solutions required more copy&pasting.

Reply:
maybe just hoist this logic into a `prepare_dataset` function that takes `args` and `num_samples`

Reply:
Refactored. The parameter list is a bit long, but I prefer that over either sending the whole train_args or using a nested function.

Reply:
looks great